kaggle_environments/envs/halite/helpers.py

```diff
@@ -0,0 +1,780 @@
+        board = deepcopy(self)
+        configuration = board.configuration
+        convert_cost = configuration.convert_cost
+        spawn_cost = configuration.spawn_cost
+        uid_counter = 0
+
+        # This is a consistent way to generate unique strings to form ship and shipyard ids
+        def create_uid():
+            nonlocal uid_counter
+            uid_counter += 1
+            return f"{self.step}-{uid_counter}"
+
+        # Process actions and store the results in the ships and shipyards lists for collision checking
+        for player in board.players.values():
+            leftover_convert_halite = 0
+
+            for shipyard in player.shipyards:
+                if shipyard.next_action == ShipyardAction.SPAWN and player.halite > spawn_cost:
+                    # Handle SPAWN actions
+                    player._halite -= spawn_cost
+                    board._add_ship(Ship(ShipId(create_uid()), shipyard.position, 0, player.id, board))
+                # Clear the shipyard's action so it doesn't repeat the same action automatically
+                shipyard.next_action = None
+
+            for ship in player.ships:
+                if ship.next_action == ShipAction.CONVERT:
+                    # Can't convert on an existing shipyard but you can use halite in a ship to fund conversion
+                    if ship.cell.shipyard_id is None and (ship.halite + player.halite) > convert_cost:
+                        # Handle CONVERT actions
+                        delta_halite = ship.halite - convert_cost
+                        # Excess halite leftover from conversion is added to the player's total only after all conversions have completed
+                        # This is to prevent the edge case of chaining halite from one convert to fund other converts
+                        leftover_convert_halite += max(delta_halite, 0)
+                        player._halite += min(delta_halite, 0)
+                        board._add_shipyard(Shipyard(ShipyardId(create_uid()), ship.position, player.id, board))
+                        board._delete_ship(ship)
+                elif ship.next_action is not None:
+                    # If the action is not None and is not CONVERT it must be NORTH, SOUTH, EAST, or WEST
+                    ship.cell._ship_id = None
+                    ship._position = wrap_point(translate_point(ship.position, ship.next_action.to_point()), configuration.size)
+                    # We don't set the new cell's ship_id here as it would be overwritten by another ship in the case of collision.
+                    # Later we'll iterate through all ships and re-set the cell._ship_id as appropriate.
+                # Clear the ship's action so it doesn't repeat the same action automatically
+                ship.next_action = None
+
+            player._halite += leftover_convert_halite
+
+        def resolve_collision(ships: List[Ship]) -> Tuple[Optional[Ship], List[Ship]]:
```

Comment:
These methods are specific to the game's rules and could be `staticmethod`? I wonder if it's worth factoring out to a separate `class` for clarity?